engine/classes/ElggCloudFilestore.php

```diff
@@ -0,0 +1,189 @@
+<?php
+use Gaufrette\Filesystem;
+
+/**
+ * A filestore that uses a Cloud storage service
+ *
+ * @warning This should be used by a wrapper class
+ * like {@link \ElggFile}.
+ */
+class ElggCloudFilestore extends \ElggFilestore {
+	/**
+	 *
+	 * @var Filesystem;
+	 */
+	private $fs;
+	
+	private $adapter;
+	
+	public function __construct($adapter = null) {
+		if ($adapter) {
+			$this->adapter = $adapter;
+			$this->fs = $adapter->getFilesystem();
```

Comment:
This class is going to blow up if it is used without fs being set. What was the motivation for nullable adapter here?